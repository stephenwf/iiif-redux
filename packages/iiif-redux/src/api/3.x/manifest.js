import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const manifest = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingDirection
   * - getBehavior
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );

  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   * - getStart
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id];
    }
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getStructures
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(getItemIds, getAllCanvases, mapAllById);

  const getCanvasIds = getItemIds;
  const getCanvases = getItems;

  const getStructureIds = createSelector(selector, structural.getStructures);
  const getStructures = createSelector(
    getStructureIds,
    getAllRanges,
    mapAllById
  );

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    getId,
    getType,
    getViewingDirection,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getItemIds,
    getItems,
    getCanvasIds,
    getCanvases,
    getStructureIds,
    getStructures,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(manifest);

export default manifest;
