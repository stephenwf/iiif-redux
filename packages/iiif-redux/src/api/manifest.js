import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import validUrl from 'valid-url';
import * as technical from './iiif-technical';
import * as descriptive from './iiif-descriptive';
import {
  getAllAnnotationLists,
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllRanges,
  getAllSequences,
  getAllServices,
} from './all';
import * as linking from './iiif-linking';
import * as structural from './iiif-structural';

export default memoize(selector => {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   * - getViewingDirection
   * - getNavDate
   **************************************************/
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingHint = createSelector(
    selector,
    technical.getWhitelistedViewingHint(['individuals', 'paged', 'continuous'])
  );

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getNavDate = createSelector(selector, technical.getNavDate);

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getDescription
   * - getMetadata
   * - getAttribution
   * - getLogo
   * - getLicence
   * - getThumbnail
   **************************************************/
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getDescription = createSelector(selector, descriptive.getDescription);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getAttribution = createSelector(selector, descriptive.getAttribution);
  const getLogo = createSelector(selector, descriptive.getLogo);

  const getLicense = createSelector(selector, descriptive.getLicense);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);
  const getThumbnail = createSelector(
    getThumbnailId,
    getAllImages,
    (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
  );

  /**************************************************
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getRelated
   * - getRendering
   * - getWithin
   **************************************************/
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    (seeAlsoIds, allExternalResources) =>
      seeAlsoIds.map(seeAlsoId => allExternalResources[seeAlsoId])
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(
    getServiceIds,
    getAllServices,
    (serviceIds, allServices) =>
      serviceIds.map(serviceId => allServices[serviceId])
  );

  const getRelatedIds = createSelector(selector, linking.getRelated);
  const getRelated = createSelector(
    getRelatedIds,
    getAllExternalResources,
    (relatedIds, allExternalResources) =>
      relatedIds.map(relatedId => allExternalResources[relatedId])
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    (renderingIds, allExternalResources) =>
      renderingIds.map(renderingId => allExternalResources[renderingId])
  );

  const getWithinIds = createSelector(selector, linking.getWithin);
  const getWithin = createSelector(
    getWithinIds,
    getAllLayers,
    getAllExternalResources,
    (withinIds, allLayers, allExternalResources) =>
      withinIds
        .map(within => allLayers[within.id] || allExternalResources[within.id])
        .filter(e => e)
  );

  /**************************************************
   * Structural properties
   *
   * - getSequences (Required)
   * - getStructures / getRanges
   **************************************************/
  const getSequenceIds = createSelector(selector, structural.getSequences);
  const getSequences = createSelector(
    getSequenceIds,
    getAllSequences,
    (sequenceIds, allSequences) =>
      sequenceIds.map(sequenceId => allSequences[sequenceId])
  );

  const getRangeIds = createSelector(selector, structural.getRanges);
  const getRanges = createSelector(
    getRangeIds,
    getAllRanges,
    (rangeIds, allRanges) => rangeIds.map(rangeId => allRanges[rangeId])
  );

  const getStructureIds = getRangeIds;
  const getStructures = getRanges;

  const getOtherContentIds = createSelector(
    selector,
    structural.getOtherContent
  );
  const getOtherContent = createSelector(
    getOtherContentIds,
    getAllExternalResources,
    getAllAnnotationLists,
    (otherContentIds, allExternalResources, allAnnotationLists) =>
      otherContentIds.map(
        otherContentId =>
          allAnnotationLists[otherContentId] ||
          allExternalResources[otherContentId] || {
            '@id': otherContentId,
            label: [{ '@value': 'unknown', '@language': '@none' }],
          }
      )
  );

  return {
    selector,
    // Technical
    getId,
    getType,
    getViewingHint,
    getViewingDirection,
    getNavDate,
    // Descriptive
    getLabel,
    getDescription,
    getMetadata,
    getAttribution,
    getLogo,
    getLicense,
    getThumbnailId,
    getThumbnail,
    // Linking
    getWithinIds,
    getWithin,
    getRenderingIds,
    getRendering,
    getRelatedIds,
    getRelated,
    getServiceIds,
    getService,
    getSeeAlsoIds,
    getSeeAlso,
    // Structural
    getSequenceIds,
    getSequences,
    getRangeIds,
    getRanges,
    getStructureIds,
    getStructures,
    getOtherContentIds,
    getOtherContent,
  };
});
