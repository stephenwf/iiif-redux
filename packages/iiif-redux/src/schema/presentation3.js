import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: 'id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations'); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollection'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReference'); // 10

// Unofficial types.
const externalResource = createEntity('externalResources');
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'externalResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    canvasReference,
    canvas,
  },
  entity => {
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBody = new schema.Union(
  {
    contentResource,
    externalResource,
    choice,
  },
  input => {
    switch (input.type) {
      case 'Application':
      case 'Dataset':
      case 'Image':
      case 'Sound':
      case 'Text':
      case 'Video':
      case 'TextualBody':
        return 'contentResource';
      case 'Choice':
        return 'choice';

      // @todo more content to add.
      default:
        return 'externalResource';
    }
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    range,
    canvasReference,
  },
  input => {
    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: annotationBody,
  target: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  rendering: [externalResource],
  partOf: partOf,
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  externalResource,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
