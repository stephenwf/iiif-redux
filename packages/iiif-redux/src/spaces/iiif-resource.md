# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
