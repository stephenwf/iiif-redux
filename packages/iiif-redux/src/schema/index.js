import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
