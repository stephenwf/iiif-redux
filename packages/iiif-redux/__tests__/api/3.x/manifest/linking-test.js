import manifest from '../../../../src/api/3.x/manifest';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/manifest/linking', () => {
  fixtures.forEach((fixture, num) => {
    describe(`Manifest number ${num}`, () => {
      const {
        result: { id },
        entities,
      } = normalize(fixture);
      const state = {
        resources: { ...DEFAULT_STATE.resources, ...entities },
      };
      const selectManifest = s => s.resources.manifests[id];

      const {
        getSeeAlso,
        getSeeAlsoIds,
        getService,
        getServiceIds,
        getLogo,
        getLogoIds,
        getHomepage,
        getHomepageId,
        getRendering,
        getRenderingIds,
        getPartOf,
        getPartOfId,
        getStartId,
        getStart,
      } = manifest(selectManifest);

      test('getSeeAlso', () => {
        expect(getSeeAlso(state)).toMatchSnapshot(`getSeeAlso ${num}`);
      });

      test('getSeeAlsoIds', () => {
        expect(getSeeAlsoIds(state)).toMatchSnapshot(`getSeeAlsoIds ${num}`);
      });

      test('getService', () => {
        expect(getService(state)).toMatchSnapshot(`getService ${num}`);
      });

      test('getServiceIds', () => {
        expect(getServiceIds(state)).toMatchSnapshot(`getServiceIds ${num}`);
      });

      test('getLogo', () => {
        expect(getLogo(state)).toMatchSnapshot(`getLogo ${num}`);
      });

      test('getLogoIds', () => {
        expect(getLogoIds(state)).toMatchSnapshot(`getLogoIds ${num}`);
      });

      test('getHomepage', () => {
        expect(getHomepage(state)).toMatchSnapshot(`getHomepage ${num}`);
      });

      test('getHomepageId', () => {
        expect(getHomepageId(state)).toMatchSnapshot(`getHomepageId ${num}`);
      });

      test('getRendering', () => {
        expect(getRendering(state)).toMatchSnapshot(`getRendering ${num}`);
      });

      test('getRenderingIds', () => {
        expect(getRenderingIds(state)).toMatchSnapshot(
          `getRenderingIds ${num}`
        );
      });

      test('getPartOf', () => {
        expect(getPartOf(state)).toMatchSnapshot(`getPartOf ${num}`);
      });

      test('getPartOfId', () => {
        expect(getPartOfId(state)).toMatchSnapshot(`getPartOfId ${num}`);
      });

      test('getStart', () => {
        expect(getStart(state)).toMatchSnapshot(`getStart ${num}`);
      });

      test('getStartId', () => {
        expect(getStartId(state)).toMatchSnapshot(`getStartId ${num}`);
      });
    });
  });
});
