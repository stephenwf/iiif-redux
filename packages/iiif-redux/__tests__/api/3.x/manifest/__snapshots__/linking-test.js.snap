// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/manifest/linking Manifest number 0 getHomepage: getHomepage 0 1`] = `
Object {
  "format": "text/html",
  "id": "https://example.org/info/book1/",
  "label": Object {
    "en": Array [
      "Home page for Book 1",
    ],
  },
  "type": "Text",
}
`;

exports[`api/3.x/manifest/linking Manifest number 0 getHomepageId: getHomepageId 0 1`] = `"https://example.org/info/book1/"`;

exports[`api/3.x/manifest/linking Manifest number 0 getLogo: getLogo 0 1`] = `
Array [
  Object {
    "id": "https://example.org/logos/institution1.jpg",
    "service": Array [
      "https://example.org/service/inst1",
    ],
    "type": "Image",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getLogoIds: getLogoIds 0 1`] = `
Array [
  "https://example.org/logos/institution1.jpg",
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getPartOf: getPartOf 0 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "type": "Collection",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getPartOfId: getPartOfId 0 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "schema": "collection",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getRendering: getRendering 0 1`] = `
Array [
  Object {
    "format": "application/pdf",
    "id": "https://example.org/iiif/book1.pdf",
    "label": Object {
      "en": Array [
        "Download as PDF",
      ],
    },
    "type": "Text",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getRenderingIds: getRenderingIds 0 1`] = `
Array [
  "https://example.org/iiif/book1.pdf",
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getSeeAlso: getSeeAlso 0 1`] = `
Array [
  Object {
    "format": "text/xml",
    "id": "https://example.org/library/catalog/book1.xml",
    "profile": "https://example.org/profiles/bibliographic",
    "type": "Dataset",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getSeeAlsoIds: getSeeAlsoIds 0 1`] = `
Array [
  "https://example.org/library/catalog/book1.xml",
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getService: getService 0 1`] = `
Array [
  Object {
    "id": "https://example.org/service/example",
    "profile": "https://example.org/docs/example-service.html",
    "type": "Service",
  },
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getServiceIds: getServiceIds 0 1`] = `
Array [
  "https://example.org/service/example",
]
`;

exports[`api/3.x/manifest/linking Manifest number 0 getStart: getStart 0 1`] = `
Object {
  "height": 1000,
  "id": "https://example.org/iiif/book1/canvas/p2",
  "items": Array [
    "https://example.org/iiif/book1/page/p2/1",
  ],
  "label": Object {
    "@none": Array [
      "p. 2",
    ],
  },
  "type": "Canvas",
  "width": 750,
}
`;

exports[`api/3.x/manifest/linking Manifest number 0 getStartId: getStartId 0 1`] = `
Object {
  "id": "https://example.org/iiif/book1/canvas/p2",
  "schema": "canvas",
}
`;

exports[`api/3.x/manifest/linking Manifest number 1 getHomepage: getHomepage 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 1 getHomepageId: getHomepageId 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 1 getLogo: getLogo 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getLogoIds: getLogoIds 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getPartOf: getPartOf 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 1 getPartOfId: getPartOfId 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 1 getRendering: getRendering 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getRenderingIds: getRenderingIds 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getSeeAlso: getSeeAlso 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getSeeAlsoIds: getSeeAlsoIds 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getService: getService 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getServiceIds: getServiceIds 1 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 1 getStart: getStart 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 1 getStartId: getStartId 1 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 2 getHomepage: getHomepage 2 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 2 getHomepageId: getHomepageId 2 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 2 getLogo: getLogo 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getLogoIds: getLogoIds 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getPartOf: getPartOf 2 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 2 getPartOfId: getPartOfId 2 1`] = `null`;

exports[`api/3.x/manifest/linking Manifest number 2 getRendering: getRendering 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getRenderingIds: getRenderingIds 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getSeeAlso: getSeeAlso 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getSeeAlsoIds: getSeeAlsoIds 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getService: getService 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getServiceIds: getServiceIds 2 1`] = `Array []`;

exports[`api/3.x/manifest/linking Manifest number 2 getStart: getStart 2 1`] = `
Object {
  "id": "https://example.org/iiif/start/2",
  "selector": Object {
    "t": 14.5,
    "type": "PointSelector",
  },
  "source": "https://example.org/iiif/book1/canvas/p1",
  "type": "SpecificResource",
}
`;

exports[`api/3.x/manifest/linking Manifest number 2 getStartId: getStartId 2 1`] = `
Object {
  "id": "https://example.org/iiif/start/2",
  "schema": "canvasReference",
}
`;
